### PR TITLE
log replSet status

### PR DESCRIPTION
**Pull request by andrewhuang**
screen shot 2015-06-04 at 5 51 45 pm

### CODE REVIEW

**Comment by nathanleiby** [comment]
looks great!

consider:
- adding test coverage https://github.com/python-diamond/Diamond/blob/master/src/collectors/mongodb/test/testmongodb.py
- document status code for health mapping

File: src/collectors/mongodb/mongodb.py [hunk 1]
```diff
@@ -177,6 +177,10 @@ def collect(self):
             self._publish_transformed(data, base_prefix)
             if str_to_bool(self.config['simple']):
                 data = self._extract_simple_data(data)
+            if str_to_bool(self.config['replica']):
+                replset_data = conn.admin.command('replSetGetStatus')
```
Comment (nathanleiby) [inline src/collectors/mongodb/mongodb.py]: how to gracefully handle if replset isn't yet initiated? might give a warning msg

File: src/collectors/mongodb/mongodb.py [hunk 2]
```diff
@@ -177,6 +177,10 @@ def collect(self):
             self._publish_transformed(data, base_prefix)
             if str_to_bool(self.config['simple']):
                 data = self._extract_simple_data(data)
+            if str_to_bool(self.config['replica']):
```
Comment (nathanleiby) [inline src/collectors/mongodb/mongodb.py:184]: I think should document `replica` in `get_default_config_help`

File: src/collectors/mongodb/mongodb.py [hunk 3]
```diff
@@ -205,6 +209,21 @@ def collect(self):
                     self._publish_dict_with_prefix(collection_stats,
                                                    collection_prefix)
 
+    def _publish_replset(self, data, base_prefix):
```
Comment (nathanleiby) [inline src/collectors/mongodb/mongodb.py:219]: could add a docstring clarifying what this does `"""Docstring"""`

**Comment by andrewhuang** [comment]
Made suggested changes, minus testing (soon). Health codes here: http://docs.mongodb.org/manual/reference/replica-states/

File: src/collectors/mongodb/mongodb.py [hunk 4]
```diff
@@ -177,6 +180,14 @@ def collect(self):
             self._publish_transformed(data, base_prefix)
             if str_to_bool(self.config['simple']):
                 data = self._extract_simple_data(data)
+            if str_to_bool(self.config['replica']):
+                try:
+                    replset_data = conn.admin.command('replSetGetStatus')
+                    self._publish_replset(replset_data, base_prefix)
+                except pymongo.errors.OperationFailure as e:
```
Comment (nathanleiby) [inline src/collectors/mongodb/mongodb.py:188]: s/e/err, since you're calling `str(err)` below

File: src/collectors/mongodb/mongodb.py [hunk 5]
```diff
@@ -65,7 +65,10 @@ def get_default_config_help(self):
             'translate_collections': 'Translate dot (.) to underscores (_)'
                                      ' in collection names.',
             'ssl': 'True to enable SSL connections to the MongoDB server.'
-                    ' Default is False'
+                    ' Default is False',
+            'replica': 'True to enable replica set logging. Reports health of'
+                       'individual nodes as well as basic aggregate stats.'
```
Comment (nathanleiby) [inline src/collectors/mongodb/mongodb.py]: needs a space after "of" or before "individual"

**Comment by andrewhuang** [comment]
Fixed previous issues. Updated to support pymongo 3.0.2.

File: src/collectors/mongodb/test/testmongodb.py [hunk 6]
```diff
@@ -133,6 +134,27 @@ def test_should_ignore_unneeded_collections(self,
...
+
+        self.connection.admin.command.assert_any_call('replSetGetStatus')
+
+    @run_only_if_pymongo_is_available
+    @patch('pymongo.MongoClient')
+    @patch.object(Collector, 'publish')
+    def test_should_fail_gracefully_if_replset_dne(self,
+                                                   publish_mock,
+                                                   connector_mock):
+        assert True
```
Comment (nathanleiby) [inline src/collectors/mongodb/test/testmongodb.py]: Remove this test if not used

File: src/collectors/mongodb/test/testmongodb.py [hunk 7]
```diff
@@ -133,6 +134,27 @@ def test_should_ignore_unneeded_collections(self,
...
+    @patch.object(Collector, 'publish')
+    def test_should_publish_replset_status_if_enabled(self,
+                                                      publish_mock,
+                                                      connector_mock):
+        data = {'more_keys': long(1), 'key': 2, 'string': 'str'}
+        self._annotate_connection(connector_mock, data)
+
+        self.collector.collect()
+
+        self.connection.admin.command.assert_any_call('replSetGetStatus')
```
Comment (nathanleiby) [inline src/collectors/mongodb/test/testmongodb.py]: Possible to test the opposite -- if replica: False then it _doesnt_ call replSetGetStatus?

(e.g. Can you modify the collector config during the test?)

**Comment by nathanleiby** [comment]
minor feedback on tests

**Comment by nathanleiby** [comment]
I'd document that pymongo must now be version 3.x

E.g. in mongodb.py as well as travis test requirements (you'll need to do this for the CI to pass when making the PR to upstream Diamond)

https://github.com/python-diamond/Diamond/search?utf8=%E2%9C%93&q=pymongo

**Comment by andrewhuang** [comment]
Added tests, removed the unused case. Currently runs on staging. Should test in prod.

**Comment by nathanleiby** [comment]
:clap:

**Pull request merged**